### ISSUE TITLE
AsNativeArray is unsafe

### ISSUE THREAD

**Issue by HalfVoxel** [andywiecko/BurstTriangulator#182]
The `AsNativeArray` extension method only pins the original array while grabbing its pointer. This means that the GC is free to free the memory while the native array is still being used. A moving GC (as used by the coreclr, which Unity will switch to at some point) would also be free to move the original array's memory at any point.

This can potentially cause crashes and memory corruption.

**Comment by andywiecko** [comment]
@HalfVoxel 
I thought the current Unity GC implementation does not move objects, so this should be quite safe.
However, I guess I will have to pin the array for safety.

**Comment by HalfVoxel** [comment]
Yes, it will not move the memory at the moment. But it may deallocate it.

**Comment by andywiecko** [comment]
Okay, now I see your point. Thank you for reporting this!
I'll have to think about a possible solution. I believe this can be fixed with proper use of `GCHandle`, but honestly, I will probably move this method to an unsafe context.

**Comment by andywiecko** [comment]
Hi @HalfVoxel,

I believe I have fixed the issue with 25c2f952f5d0636acef569b037b515a4db1fda68. There are now two extensions available: 

- `AsNativeArray(out GCHandle)`
- `UnsafeAsNativeArray()`. 

However, I tried various setups and was unable to crash Unity or made memory leaks using the previous approach.

I hope this helps!

Best,
Andrzej

**Comment by HalfVoxel** [comment]
I have replicated memory corruption using `UnsafeAsNativeArray`. It's quite easy to do, but I didn't save the code, I'm afraid.

While it is technically-okay-ish to use with mono, as long as one is really careful. It's not possible to use correctly at all with coreclr. I feel like the inclusion of it would just be a ticking time bomb that will cause random memory corruption in the future for potential users who install this package and keep using the current version for a long time. I see no need for it that would be justified.

Also. Your AsNativeArray is kinda suboptimal as it still allocates a new GCHandle every time. You can use UnsafeUtility.PinGCArrayAndGetDataAddress instead, which uses cached handles, and is a fair bit faster.

**Comment by andywiecko** [comment]
Okay, I guess I'll revert this change and instead add an example note in the manual on how to achieve this using Collections' `UnsafeUtility`. This will ensure safety in the future versions. I also hope that Unity will transition to CoreCLR soon!

**Comment by HalfVoxel** [comment]
I think the best way is to have only AsNativeArray, but implement it something like this:

```
struct Handle {
    ulong gcHandle;
    public void Free () {
        UnsafeUtility.ReleaseGCObject(gcHandle);
    }
}

public static Handle AsNativeArray (T[] data, out Handle handle) {
        unsafe {
            // In Unity 2023.2+ pointers are not required, one can use Span<T> instead.
            var ptr = UnsafeUtility.PinGCArrayAndGetDataAddress(data, out var gcHandle);
            var ret = NativeArrayUnsafeUtility.ConvertExistingDataToNativeArray<T>(ptr, array.Length, Allocator.None);
#if ENABLE_UNITY_COLLECTIONS_CHECKS
            var m_SafetyHandle = AtomicSafetyHandle.Create();
            NativeArrayUnsafeUtility.SetAtomicSafetyHandle(ref ret, m_SafetyHandle);
#endif
            handle = new Handle { gcHandle = gcHandle };
            return ret;
       }
 }
 ```

**Comment by andywiecko** [comment]
I believe this is the best solution, as it provides ease of use and ensures safety.

**Comment by andywiecko** [comment]
Resolved with [v3.2.0](https://github.com/andywiecko/BurstTriangulator/releases/tag/v3.2.0)
